mypy/checker.py

```diff
@@ -7519,6 +7556,55 @@ def has_valid_attribute(self, typ: Type, name: str) -> bool:
     def get_expression_type(self, node: Expression, type_context: Type | None = None) -> Type:
         return self.expr_checker.accept(node, type_context=type_context)
 
+    def get_deprecation_warning(self, decorators: Iterable[Expression]) -> str | None:
+        for decorator in decorators:
+            if (
+                isinstance(decorator, CallExpr)
+                and isinstance(callee := decorator.callee, NameExpr)
+                and (
+                    (callee.fullname in DEPRECATED_TYPE_NAMES)
+                    or (not callee.fullname and callee.name == "deprecated")
+                )
+                and (len(args := decorator.args) == 1)
+                and isinstance(arg := args[0], StrExpr)
+            ):
+                return arg.value
+        return None
+
+    def check_deprecated_function(self, typ: Type, context: Context, memberaccess: bool) -> None:
+        if isinstance(typ := get_proper_type(typ), (CallableType, Overloaded)):
+            self._check_deprecated(typ, context, memberaccess)
+
+    def check_deprecated_class(self, typ: TypeInfo, context: Context, memberaccess: bool) -> None:
+        self._check_deprecated(typ, context, memberaccess)
+
+    def _check_deprecated(
+        self, typ: CallableType | Overloaded | TypeInfo, context: Context, memberaccess: bool
+    ) -> None:
+        if (depr := typ.deprecated) is not None:
+            if memberaccess:
+                self.warn_deprecated(typ, depr, context)
+            else:
+                for imp in self.tree.imports:
+                    if isinstance(imp, ImportFrom) and any(typ.name == n[0] for n in imp.names):
+                        break
+                else:
+                    self.warn_deprecated(typ, depr, context)
+
+    def warn_deprecated(
+        self, type_: CallableType | Overloaded | TypeInfo, deprecated: str, context: Context
+    ) -> None:
+        name = type_.name
+        if isinstance(type_, CallableType):
+            if (defn := type_.definition) is not None:
+                name = defn.fullname
+        elif isinstance(type_, Overloaded):
+            if isinstance(func := type_.items[0].definition, FuncDef):
+                name = func.fullname
+        else:
+            name = type_.fullname
+        self.msg.fail(f"{name} is deprecated - {deprecated}", context, code=codes.DEPRECATED)
```

Comment:
A dash doesn't look great here. For reference, in pyanalyze I use a colon:

```
Literal[<function show at 0x105723ba0>] is deprecated: Use 'print' instead (code: deprecated)
```

Pyright gives the user message on a separate line:

```
  /Users/jelle/Desktop/depr.py:10:1 - error: The function "show" is deprecated
    Use 'print' instead (reportDeprecated)
```

I think the colon version works best with mypy's general style of outputting messages.
```suggestion
        self.msg.fail(f"{name} is deprecated: {deprecated}", context, code=codes.DEPRECATED)
```

Reply:
okay, `warn_deprecated` now writes ": " instead of " - "